### ISSUE TITLE
net.send_config_set failed

### ISSUE THREAD

**Issue by ciscorat** [ktbyers/netmiko#1592]
Hi,

I'm using netmiko 3.0.0 with pytnon 3.8.1.
The device type:
"device_type": "cisco_nxos",

when I use net.send_config_set I get the failed to exit so I added:
		print(net.send_config_set("logging monitor 7", exit_config_mode=False))

However, it seems like base_connection.py doesn't have the "output" set:

Traceback (most recent call last):
  File "PycharmProjects/misc/XXXXXXXXXX.py", line 144, in <module>
    print(net.send_config_set("logging monitor 7", exit_config_mode=False))
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/netmiko/base_connection.py", line 1740, in send_config_set
    output += new_output
UnboundLocalError: local variable 'output' referenced before assignment

I added the following line just for debug purpose:

        """
        delay_factor = self.select_delay_factor(delay_factor)
        output = ""
       ~~~~~~~~~~
        if config_commands is None:
            return ""
        elif isinstance(config_commands, str):
            config_commands = (config_commands,)

So now it doesn't fail but the debug shows that the config command doesn't go into the config mode:

EBUG:netmiko:read_channel:
DEBUG:netmiko:[find_prompt()]: prompt is AY03-eNSE-9508-02#
<switch name hidden># <<<<<<<<<<<<<<<<<<<<<<<<<<<<prompt>>>>>>>>>>>>>>>>>>>>>>
DEBUG:netmiko:write_channel: b'\n'
DEBUG:netmiko:read_channel:
<switch name hidden>#
DEBUG:netmiko:read_channel:
DEBUG:netmiko:read_channel:
DEBUG:netmiko:write_channel: b'logging monitor 7\n'
DEBUG:netmiko:Pattern is: logging\ monitor\ 7
DEBUG:netmiko:_read_channel_expect read_data: logging monitor 7

DEBUG:netmiko:Pattern found: logging\ monitor\ 7 logging monitor 7

DEBUG:netmiko:Pattern is: (?:<switch name hidden>#)
DEBUG:netmiko:_read_channel_expect read_data:                      ^
% Invalid command at '^' marker.
XXXXXX-9508-02#
DEBUG:netmiko:Pattern found: (?:<switch name hidden>#)                      ^
% Invalid command at '^' marker.
<switch name hidden>#
DEBUG:netmiko:logging monitor 7
                     ^
% Invalid command at '^' marker.

<switch name hidden>#
logging monitor 7
                     ^
% Invalid command at '^' marker.

<switch name hidden>#
YDEKEL-M-D6YG:~ XXXXX$
 
I have added that just in case:
		print(net.send_config_set("logging monitor 7", exit_config_mode=False, enter_config_mode=True))

Basically that what I'm doing:
try:
	net = BaseConnection(**lab)
	print(f"""Connected to {lab["ip"]}""")
except:
    print(f"""Connect to {lab["ip"]} failed""")
    exit()

if debug:
	print(net.find_prompt())

	try:
		output = ""
		print(net.find_prompt(), "<<<<<<<<<<<<<<<<<<<<<<<<<<<<prompt>>>>>>>>>>>>>>>>>>>>>>")
		print(net.send_config_set("logging monitor 7", exit_config_mode=False, enter_config_mode=True))
	except Exception as e:
		print ("Exception: {}".format(e), "\n++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++")
		raise
		
	# print(net.send_command("show boot"))

	exit()

Can you please help me figure this out?

**Comment by ktbyers** [comment]
I think that issue was fixed here:

https://github.com/ktbyers/netmiko/pull/1557

You will need to use the `develop` branch until I release a new version of Netmiko (or possibly Netmiko 2.4.2).

**Comment by ciscorat** [comment]
I have applied the fix in base connection but it doesn't enter config mode:
        # Send config commands
        output = ""
        if enter_config_mode:
            cfg_mode_args = (config_mode_command,) if config_mode_command else tuple()
            output += self.config_mode(*cfg_mode_args)



 File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/netmiko/base_connection.py", line 1753, in send_config_set
    output += self.exit_config_mode()

**Comment by ktbyers** [comment]
Please post the full stack trace that you get with the patch applied.

**Comment by ciscorat** [comment]
Traceback (most recent call last):
  File "PycharmProjects/misc/eNSE-code-up-downgrade.py", line 143, in <module>
    print(net.send_config_set("logging monitor 7"))
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/netmiko/base_connection.py", line 1753, in send_config_set
    output += self.exit_config_mode()
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/netmiko/base_connection.py", line 1635, in exit_config_mode
    raise ValueError("Failed to exit configuration mode")
ValueError: Failed to exit configuration mode

The "patch" is just the code change as reflected here:
https://github.com/ktbyers/netmiko/pull/1557/files

        if not hasattr(config_commands, "__iter__"):
            raise ValueError("Invalid argument passed into send_config_set")

        # Send config commands
        output = ""
        if enter_config_mode:
            cfg_mode_args = (config_mode_command,) if config_mode_command else tuple()
            output += self.config_mode(*cfg_mode_args)

        if self.fast_cli:

**Comment by ciscorat** [comment]
P.S when I disable exit config it doesn't raise exception but it still doesn't enter the config mode:

DEBUG:netmiko:_read_channel_expect read_data: logging monitor 7

DEBUG:netmiko:Pattern found: logging\ monitor\ 7 logging monitor 7

DEBUG:netmiko:Pattern is: (?:<switch name>|#)
DEBUG:netmiko:_read_channel_expect read_data:                      ^
% Invalid command at '^' marker.
<switch name>#
DEBUG:netmiko:Pattern found: (?:<switch name>|#)                      ^
% Invalid command at '^' marker.
<switch name>#
DEBUG:netmiko:logging monitor 7
                     ^
% Invalid command at '^' marker.

<switch name>#
logging monitor 7
                     ^
% Invalid command at '^' marker.

<switch name>#

**Comment by ktbyers** [comment]
Can you include the full debug log?

**Comment by ciscorat** [comment]
DEBUG:paramiko.transport:starting thread (client mode): 0x1fd6ac0
DEBUG:paramiko.transport:Local version/idstring: SSH-2.0-paramiko_2.7.1
DEBUG:paramiko.transport:Remote version/idstring: SSH-2.0-OpenSSH_7.2 FIPS
INFO:paramiko.transport:Connected (version 2.0, client OpenSSH_7.2)
DEBUG:paramiko.transport:kex algos:['curve25519-sha256@libssh.org', 'ecdh-sha2-nistp256', 'ecdh-sha2-nistp384', 'diffie-hellman-group-exchange-sha256', 'diffie-hellman-group14-sha1'] server key:['ssh-rsa'] client encrypt:['aes128-ctr', 'aes192-ctr', 'aes256-ctr'] server encrypt:['aes128-ctr', 'aes192-ctr', 'aes256-ctr'] client mac:['hmac-sha1'] server mac:['hmac-sha1'] client compress:['none', 'zlib@openssh.com'] server compress:['none', 'zlib@openssh.com'] client lang:[''] server lang:[''] kex follows?False
DEBUG:paramiko.transport:Kex agreed: curve25519-sha256@libssh.org
DEBUG:paramiko.transport:HostKey agreed: ssh-rsa
DEBUG:paramiko.transport:Cipher agreed: aes128-ctr
DEBUG:paramiko.transport:MAC agreed: hmac-sha1
DEBUG:paramiko.transport:Compression agreed: none
DEBUG:paramiko.transport:kex engine KexCurve25519 specified hash_algo <built-in function openssl_sha256>
DEBUG:paramiko.transport:Switch to new keys ...
DEBUG:paramiko.transport:Adding ssh-rsa host key for <switch ip hidden>: b'6ad7129a4f625ab27254074fe3cf02a8'
DEBUG:paramiko.transport:userauth is OK
INFO:paramiko.transport:Auth banner: b'\n***************************************************************************\n                            NOTICE TO USERS\nThis is a private computer system and is for authorized use only. Users\n(authorized or unauthorized) have no explicit or implicit expectation of\nprivacy.\n\nAny or all uses of this system and all files on this system may be\nintercepted, monitored, recorded, copied, audited, inspected, and disclosed\nto authorized site and law enforcement personnel, as well as authorized\nofficials of other agencies, both domestic and foreign. By using this\nsystem, the user consents to such interception, monitoring, recording,\ncopying, auditing, inspection, and disclosure at the discretion of the\nauthorized site or personnel.\n\nUnauthorized or improper use of this system may result in administrative\ndisciplinary action and civil and criminal penalties. By continuing to\nuse this system you indicate your awareness of and consent to these terms\nand conditions of use. LOG OFF IMMEDIATELY if you do not agree to the\nconditions stated in this warning.\n*****************************************************************************\n'
INFO:paramiko.transport:Authentication (password) successful!
DEBUG:paramiko.transport:[chan 0] Max packet in: 32768 bytes
DEBUG:paramiko.transport:Received global request "hostkeys-00@openssh.com"
DEBUG:paramiko.transport:Rejecting "hostkeys-00@openssh.com" global request from server.
DEBUG:paramiko.transport:[chan 0] Max packet out: 32768 bytes
DEBUG:paramiko.transport:Secsh channel 0 opened.
DEBUG:paramiko.transport:[chan 0] Sesch channel 0 request ok
DEBUG:paramiko.transport:[chan 0] Sesch channel 0 request ok
DEBUG:netmiko:read_channel:
Cisco Nexus Operating System (NX-OS) Software
TAC support: http://www.cisco.com/tac
Copyright (C) 2002-2017, Cisco and/or its affiliates.
All rights reserved.
The copyrights to certain works contained in this software are
owned by other third parties and used and distributed under their own
licenses, such as open source.  This software is provided "as is," and unless
otherwise stated, there is no warranty, express or implied, including but not
limited to warranties of merchantability and fitness for a particular purpose.
Certain components of this software are licensed under
the GNU General Public License (GPL) version 2.0 or
GNU General Public License (GPL) version 3.0  or the GNU
Lesser General Public License (LGPL) Version 2.1 or
Lesser General Public License (LGPL) Version 2.0.
A copy of each such license is available at
http://www.opensource.org/licenses/gpl-2.0.php and
http://opensource.org/licenses/gpl-3.0.html and
http://www.opensource.org/licenses/lgpl-2.1.php and
http://www.gnu.org/licenses/old-licenses/library.txt.
<switch name hidden>#
DEBUG:netmiko:read_channel:
DEBUG:netmiko:read_channel:
DEBUG:netmiko:read_channel:
DEBUG:netmiko:write_channel: b'\n'
DEBUG:netmiko:read_channel:
<switch name hidden>#
DEBUG:netmiko:read_channel:
DEBUG:netmiko:[find_prompt()]: prompt is <switch name hidden>#
DEBUG:netmiko:read_channel:
DEBUG:netmiko:In disable_paging
DEBUG:netmiko:Command: terminal length 0

DEBUG:netmiko:write_channel: b'terminal length 0\n'
DEBUG:netmiko:Pattern is: terminal\ length\ 0
DEBUG:netmiko:_read_channel_expect read_data: terminal length 0

DEBUG:netmiko:Pattern found: terminal\ length\ 0 terminal length 0

DEBUG:netmiko:terminal length 0

DEBUG:netmiko:Exiting disable_paging
<switch name hidden># channel:
DEBUG:netmiko:Clear buffer detects data in the channel
DEBUG:netmiko:read_channel:
Connected to <switch ip hidden>
DEBUG:netmiko:read_channel:
DEBUG:netmiko:write_channel: b'\n'
DEBUG:netmiko:read_channel:
<switch name hidden>#
DEBUG:netmiko:read_channel:
DEBUG:netmiko:[find_prompt()]: prompt is <switch name hidden>#
<switch name hidden>#
DEBUG:netmiko:write_channel: b'\n'
DEBUG:netmiko:read_channel:
<switch name hidden>#
DEBUG:netmiko:read_channel:
DEBUG:netmiko:read_channel:
DEBUG:netmiko:write_channel: b'logging monitor 7\n'
DEBUG:netmiko:Pattern is: logging\ monitor\ 7
DEBUG:netmiko:_read_channel_expect read_data: logging monitor 7

DEBUG:netmiko:Pattern found: logging\ monitor\ 7 logging monitor 7

DEBUG:netmiko:Pattern is: (?:XXXX\-XXXX\-XXXX\-XX|#)
DEBUG:netmiko:_read_channel_expect read_data:                      ^
% Invalid command at '^' marker.
<switch name hidden>#
DEBUG:netmiko:Pattern found: (?:XXXX\-XXXX\-XXXX\-XX|#)                      ^
% Invalid command at '^' marker.
<switch name hidden>#
DEBUG:netmiko:logging monitor 7
                     ^
% Invalid command at '^' marker.

<switch name hidden>#
logging monitor 7
                     ^
% Invalid command at '^' marker.

<switch name hidden>#

**Comment by ciscorat** [comment]
I added the following to base_connection:
        if enter_config_mode:
            log.debug("$$$$$$$$$$$$$$$In config mode$$$$$$$$$$$$$$$$$$$$")
            cfg_mode_args = (config_mode_command,) if config_mode_command else tuple()
            output += self.config_mode(*cfg_mode_args)
        else:
        	log.debug("!!!!!!!!!!!!!!!failed to enter config mode!!!!!!!!!!!!!!")

I confirm it does enter config mode:

DEBUG:netmiko:$$$$$$$$$$$$$$$In config mode$$$$$$$$$$$$$$$$$$$$
DEBUG:netmiko:write_channel: b'\n'
DEBUG:netmiko:read_channel:
AY03-eNSE-9508-02#
DEBUG:netmiko:read_channel:
DEBUG:netmiko:read_channel:
DEBUG:netmiko:write_channel: b'logging monitor 7\n'
DEBUG:netmiko:Pattern is: logging\ monitor\ 7
DEBUG:netmiko:_read_channel_expect read_data: logging monitor 7

DEBUG:netmiko:Pattern found: logging\ monitor\ 7 logging monitor 7

DEBUG:netmiko:Pattern is: (?:AY03\-eNSE\-9508\-02|#)

So the problem relies maybe here:
`DEBUG:netmiko:Pattern found: logging\ monitor\ 7 logging monitor 7`
Why the command displayed twice?
Look at line 559:
' log.debug(f"Pattern found: {pattern} {output}")'

Does this make sense that the `pattern` and the `output` match?
'DEBUG:netmiko:Pattern found: logging\ monitor\ 7'
'DEBUG:netmiko:Pattern is: (?:AY03\-eNSE\-9508\-02|#)'

**Comment by ciscorat** [comment]
Hey Kirk,

I just checked with XR and it has the same problem so I used another netmiko(2.2.2) on a colleague's laptop with the same XR. It clear doesn't go into command mode:
DEBUG:netmiko:write_channel: b'config term\n'
DEBUG:netmiko:Pattern is: RP/0/RSP1/CPU0:9
DEBUG:netmiko:_read_channel_expect read_data: config term

DEBUG:netmiko:_read_channel_expect read_data: Tue Feb 25 14:29:48.756 UTC

DEBUG:netmiko:_read_channel_expect read_data: RP/0/RSP1/CPU0:9910(config)#
DEBUG:netmiko:Pattern found: RP/0/RSP1/CPU0:9 config term
Tue Feb 25 14:29:48.756 UTC
RP/0/RSP1/CPU0:9910(config)#

As opposed to mine:

DEBUG:netmiko:read_channel: Tue Feb 25 14:06:51.228 UTC
RP/0/RSP1/CPU0:9910#
DEBUG:netmiko:Clear buffer detects data in the channel
DEBUG:netmiko:read_channel:
DEBUG:netmiko:write_channel: b'\n'
DEBUG:netmiko:read_channel:
RP/0/RSP1/CPU0:9910#
DEBUG:netmiko:read_channel:
DEBUG:netmiko:read_channel:
DEBUG:netmiko:write_channel: b'logging buffered debugging\n'
DEBUG:netmiko:Pattern is: logging\ buffered\ debugging
DEBUG:netmiko:_read_channel_expect read_data: logging buffered debugging

DEBUG:netmiko:Pattern found: logging\ buffered\ debugging logging buffered debugging

DEBUG:netmiko:Pattern is: (?:RP/0/RSP1/CPU0:9910|#)
DEBUG:netmiko:_read_channel_expect read_data:                          ^
% Invalid input detected at '^' marker.
RP/0/RSP1/CPU0:9910#

**Comment by ciscorat** [comment]
The code for above is pretty basic:

#!/usr/bin/env python3
from cec import cec
from netmiko.base_connection import BaseConnection
from os import system
import os,logging, sys

_ = system('clear') 

debug = True
if debug:
	logging.basicConfig(level=logging.DEBUG)
	logger = logging.getLogger("netmiko")

lab = {
	"device_type": "cisco_nxos",
    "username": os.environ['USER'],
    "ip": sys.argv[1],
    "password": (cec()),
    'banner_timeout': 12,
    'global_delay_factor': 3,
}
net_connect= BaseConnection(**lab)
net_connect.send_config_set("logging buffered debugging")

**Comment by ktbyers** [comment]
@ciscorat Is your device's prompt really just a `#` (and nothing else)?

That very likely won't work.


Kirk

**Comment by ciscorat** [comment]
I hid the entire prompt of the NX however. I did show the prompt of the XR:
ASR9010

DEBUG:netmiko:read_channel: Tue Feb 25 14:06:51.228 UTC
RP/0/RSP1/CPU0:9910#
DEBUG:netmiko:Clear buffer detects data in the channel
DEBUG:netmiko:read_channel:
DEBUG:netmiko:write_channel: b'\n'
DEBUG:netmiko:read_channel:
RP/0/RSP1/CPU0:9910#
DEBUG:netmiko:read_channel:
DEBUG:netmiko:read_channel:
DEBUG:netmiko:write_channel: b'logging buffered debugging\n'
DEBUG:netmiko:Pattern is: logging\ buffered\ debugging
DEBUG:netmiko:_read_channel_expect read_data: logging buffered debugging

DEBUG:netmiko:Pattern found: logging\ buffered\ debugging logging buffered debugging

DEBUG:netmiko:Pattern is: (?:RP/0/RSP1/CPU0:9910|#)
DEBUG:netmiko:_read_channel_expect read_data: ^
% Invalid input detected at '^' marker.
RP/0/RSP1/CPU0:9910#

so I used another netmiko(2.2.2) on a colleague's laptop with the same XR device. Its clear doesn't go into command mode with netmiko 3.0.0 but it does with netmiko 2.2.2:

DEBUG:netmiko:write_channel: b'config term\n'
DEBUG:netmiko:Pattern is: RP/0/RSP1/CPU0:9
DEBUG:netmiko:_read_channel_expect read_data: config term

DEBUG:netmiko:_read_channel_expect read_data: Tue Feb 25 14:29:48.756 UTC

DEBUG:netmiko:_read_channel_expect read_data: RP/0/RSP1/CPU0:9910(config)#
DEBUG:netmiko:Pattern found: RP/0/RSP1/CPU0:9 config term
Tue Feb 25 14:29:48.756 UTC
RP/0/RSP1/CPU0:9910(config)#

**Comment by ciscorat** [comment]
I just added the following to base_connection.py:

        :param enter_config_mode: Do you enter config mode before sending config commands
        :type exit_config_mode: bool

        """
        log.debug(f"netmiko:additional debug: ----------------------------------in send_config_set")
        delay_factor = self.select_delay_factor(delay_factor)
        if config_commands is None:
            return ""
        elif isinstance(config_commands, str):
            config_commands = (config_commands,)

seems like it doesn't write `config t` to the channel:

DEBUG:netmiko:netmiko:additional debug: ----------------------------------in send_config_set
DEBUG:netmiko:write_channel: b'\n'
DEBUG:netmiko:read_channel:
RP/0/RSP1/CPU0:9910-A#
DEBUG:netmiko:read_channel:
DEBUG:netmiko:read_channel:
DEBUG:netmiko:write_channel: b'logging buffered debugging\n'
DEBUG:netmiko:Pattern is: logging\ buffered\ debugging
DEBUG:netmiko:_read_channel_expect read_data: logging buffered debugging

**Comment by ktbyers** [comment]
Okay, one problem at a time...which one do you want to look at XR or NX-OS?

**Comment by ciscorat** [comment]
seems like they  both don't send 'config t' so this is beyond just one device type.

**Comment by ktbyers** [comment]
Okay, I am going to work on the NX-OS issue. I am somewhat doubtful they are the same issue, but if they are we will find that out in the process.

Can you post the code associated with the NX-OS logging output (at least the relevant section)?

**Comment by ciscorat** [comment]
Here is the debug log from the NX which as well doesn't send the 'conf t':
DEBUG:netmiko:netmiko:additional debug: ----------------------------------in send_config_set
DEBUG:netmiko:write_channel: b'\n'
DEBUG:netmiko:read_channel:
<switch name hidden>#
DEBUG:netmiko:read_channel:
DEBUG:netmiko:read_channel:
DEBUG:netmiko:write_channel: b'logging buffered debugging\n'

**Comment by ciscorat** [comment]
more info
I added the following to cisco_base_connection.py:
        Cisco IOS devices abbreviate the prompt at 20 chars in config mode
        """
        log.debug(f"netmiko:additional debug: ----------------------------------send config term")
        if not pattern:
            pattern = re.escape(self.base_prompt[:16])
        return super().config_mode(config_command=config_command, pattern=pattern)

That's the output, never enters `config term`
DEBUG:netmiko:netmiko:additional debug: ----------------------------------in send_config_set
DEBUG:netmiko:write_channel: b'\n'
DEBUG:netmiko:read_channel:
[switch name hidden]#
DEBUG:netmiko:read_channel:
DEBUG:netmiko:read_channel:
DEBUG:netmiko:write_channel: b'logging buffered debugging\n'
DEBUG:netmiko:Pattern is: logging\ buffered\ debugging

sends the config command into the exec mode

**Comment by ciscorat** [comment]
Entire code and log:

#!/usr/bin/env python3
from cec import cec
from netmiko.base_connection import BaseConnection
from os import system
import os,logging, sys

_ = system('clear') 

debug = True
if debug:
	logging.basicConfig(level=logging.DEBUG)
	logger = logging.getLogger("netmiko")

lab = {
	"device_type": "cisco_xr",
    "username": os.environ['USER'],
    "ip": sys.argv[1],
    "ssh_config_file": ".ssh/config",
    'banner_timeout': 12,
    "password": (cec()),
    'banner_timeout': 12,
    'global_delay_factor': 3,
}
net_connect= BaseConnection(**lab)
net_connect.send_config_set("logging buffered debugging", exit_config_mode=False, delay_factor=100)





YDEKEL-M-D6YG:~$ python3 PycharmProjects/misc/test-xr.py 9.0.6.202
DEBUG:paramiko.transport:starting thread (client mode): 0x8742ee0
DEBUG:paramiko.transport:Local version/idstring: SSH-2.0-paramiko_2.7.1
DEBUG:paramiko.transport:Remote version/idstring: SSH-2.0-OpenSSH_7.2 FIPS
INFO:paramiko.transport:Connected (version 2.0, client OpenSSH_7.2)
DEBUG:paramiko.transport:kex algos:['curve25519-sha256@libssh.org', 'ecdh-sha2-nistp256', 'ecdh-sha2-nistp384', 'diffie-hellman-group-exchange-sha256', 'diffie-hellman-group14-sha1'] server key:['ssh-rsa'] client encrypt:['aes128-ctr', 'aes192-ctr', 'aes256-ctr'] server encrypt:['aes128-ctr', 'aes192-ctr', 'aes256-ctr'] client mac:['hmac-sha1'] server mac:['hmac-sha1'] client compress:['none', 'zlib@openssh.com'] server compress:['none', 'zlib@openssh.com'] client lang:[''] server lang:[''] kex follows?False
DEBUG:paramiko.transport:Kex agreed: curve25519-sha256@libssh.org
DEBUG:paramiko.transport:HostKey agreed: ssh-rsa
DEBUG:paramiko.transport:Cipher agreed: aes128-ctr
DEBUG:paramiko.transport:MAC agreed: hmac-sha1
DEBUG:paramiko.transport:Compression agreed: none
DEBUG:paramiko.transport:kex engine KexCurve25519 specified hash_algo <built-in function openssl_sha256>
DEBUG:paramiko.transport:Switch to new keys ...
DEBUG:paramiko.transport:Adding ssh-rsa host key for 9.0.6.202: b'6ad7129a4f625ab27254074fe3cf02a8'
DEBUG:paramiko.transport:userauth is OK
INFO:paramiko.transport:Auth banner: b'\n***************************************************************************\n                            NOTICE TO USERS\nThis is a private computer system and is for authorized use only. Users\n(authorized or unauthorized) have no explicit or implicit expectation of\nprivacy.\n\nAny or all uses of this system and all files on this system may be\nintercepted, monitored, recorded, copied, audited, inspected, and disclosed\nto authorized site and law enforcement personnel, as well as authorized\nofficials of other agencies, both domestic and foreign. By using this\nsystem, the user consents to such interception, monitoring, recording,\ncopying, auditing, inspection, and disclosure at the discretion of the\nauthorized site or personnel.\n\nUnauthorized or improper use of this system may result in administrative\ndisciplinary action and civil and criminal penalties. By continuing to\nuse this system you indicate your awareness of and consent to these terms\nand conditions of use. LOG OFF IMMEDIATELY if you do not agree to the\nconditions stated in this warning.\n*****************************************************************************\n'
INFO:paramiko.transport:Authentication (password) successful!
DEBUG:paramiko.transport:[chan 0] Max packet in: 32768 bytes
DEBUG:paramiko.transport:Received global request "hostkeys-00@openssh.com"
DEBUG:paramiko.transport:Rejecting "hostkeys-00@openssh.com" global request from server.
DEBUG:paramiko.transport:[chan 0] Max packet out: 32768 bytes
DEBUG:paramiko.transport:Secsh channel 0 opened.
DEBUG:paramiko.transport:[chan 0] Sesch channel 0 request ok
DEBUG:paramiko.transport:[chan 0] Sesch channel 0 request ok
DEBUG:netmiko:read_channel:
Cisco Nexus Operating System (NX-OS) Software
TAC support: http://www.cisco.com/tac
Copyright (C) 2002-2017, Cisco and/or its affiliates.
All rights reserved.
The copyrights to certain works contained in this software are
owned by other third parties and used and distributed under their own
licenses, such as open source.  This software is provided "as is," and unless
otherwise stated, there is no warranty, express or implied, including but not
limited to warranties of merchantability and fitness for a particular purpose.
Certain components of this software are licensed under
the GNU General Public License (GPL) version 2.0 or
GNU General Public License (GPL) version 3.0  or the GNU
Lesser General Public License (LGPL) Version 2.1 or
Lesser General Public License (LGPL) Version 2.0.
A copy of each such license is available at
http://www.opensource.org/licenses/gpl-2.0.php and
http://opensource.org/licenses/gpl-3.0.html and
http://www.opensource.org/licenses/lgpl-2.1.php and
http://www.gnu.org/licenses/old-licenses/library.txt.
[switch name hidden]#
DEBUG:netmiko:read_channel:
DEBUG:netmiko:read_channel:
DEBUG:netmiko:read_channel:
DEBUG:netmiko:write_channel: b'\n'
DEBUG:netmiko:read_channel:
[switch name hidden]#
DEBUG:netmiko:read_channel:
DEBUG:netmiko:[find_prompt()]: prompt is [switch name hidden]#
DEBUG:netmiko:read_channel:
DEBUG:netmiko:In disable_paging
DEBUG:netmiko:Command: terminal length 0

DEBUG:netmiko:write_channel: b'terminal length 0\n'
DEBUG:netmiko:Pattern is: terminal\ length\ 0
DEBUG:netmiko:_read_channel_expect read_data: terminal length 0

DEBUG:netmiko:Pattern found: terminal\ length\ 0 terminal length 0

DEBUG:netmiko:terminal length 0

DEBUG:netmiko:Exiting disable_paging
[switch name hidden]# channel:
DEBUG:netmiko:Clear buffer detects data in the channel
DEBUG:netmiko:read_channel:
DEBUG:netmiko:netmiko:additional debug: ----------------------------------in send_config_set
DEBUG:netmiko:write_channel: b'\n'
DEBUG:netmiko:read_channel:
[switch name hidden]#
DEBUG:netmiko:read_channel:
DEBUG:netmiko:read_channel:
DEBUG:netmiko:write_channel: b'logging buffered debugging\n'
DEBUG:netmiko:Pattern is: logging\ buffered\ debugging
DEBUG:netmiko:_read_channel_expect read_data: logging buffered debugging

DEBUG:netmiko:Pattern found: logging\ buffered\ debugging logging buffered debugging

DEBUG:netmiko:Pattern is: (?:[switch name hidden]|#)
DEBUG:netmiko:_read_channel_expect read_data:                      ^
% Invalid command at '^' marker.
[switch name hidden]#
DEBUG:netmiko:Pattern found: (?:[switch name hidden]|#)                      ^
% Invalid command at '^' marker.
[switch name hidden]#
DEBUG:netmiko:logging buffered debugging
                     ^
% Invalid command at '^' marker.

[switch name hidden]#
YDEKEL-M-D6YG:~ $

**Comment by ktbyers** [comment]
Can you post the code associated with the NX-OS logging output (at least the relevant section)?

**Comment by ktbyers** [comment]
What are you doing here?

```
from netmiko.base_connection import BaseConnection
...
net_connect= BaseConnection(**lab)
```

...that is not a valid Netmiko pattern. 

You need to use `ConnectHandler`

**Comment by ktbyers** [comment]
So it should be:

```python
net_connect = ConnectHandler(**lab)
```

**Comment by ktbyers** [comment]
If you don't use `ConnectHandler`, then you will not get the correct Netmiko class (which will cause the issues that you are seeing with not entering configuration mode).

**Comment by ciscorat** [comment]
OK this is working. The reason I used BaseConnection is I had an error importing ConnectHandler.
Thanks a lot and sorry if I wasted your time
